Wrappers/Python/cil/processors/PaganinProcessor.py

```diff
@@ -204,101 +205,78 @@ def check_input(self, data):
+            data.array = np.expand_dims(data.array, axis=1) 
+            if out.array.ndim < data.array.ndim:
+                out.array = np.expand_dims(out.array, axis=1)
+
+        if data.array.ndim != 4:
             raise(ValueError('Data must be 2D or 3D per channel'))
         
-        if len(out.shape) == 2:
-            out.array = np.expand_dims(out.array, len(out.shape))
-        
         # create a filter based on the shape of the data
-        filter_shape = np.shape(data_proj)
-        self.filter_Nx = filter_shape[0]+self.pad*2
-        self.filter_Ny = filter_shape[1]+self.pad*2
+        proj_shape = data.array.shape[-2:]
+        self.filter_Nx = proj_shape[0]+self.pad*2
+        self.filter_Ny = proj_shape[1]+self.pad*2
         self._create_filter(self.override_filter)
 
+        # allocate padded buffer
+        padded_buffer = np.zeros((self.filter_Nx, self.filter_Ny), dtype=data.dtype)
+        buffer_slice = (slice(self.pad, self.pad + proj_shape[0]), 
+                    slice(self.pad, self.pad + proj_shape[1]))
+        
         # pre-calculate the scaling factor
         scaling_factor = -(1/self.mu)
-
-        # allocate padded buffer
-        padded_buffer = np.zeros(tuple(x+self.pad*2 for x in data_proj.shape), dtype=data.dtype)
+        mag2 = self.magnification ** 2
         
-        # make slice indices to unpad the data
-        if self.pad>0:
-            slice_pad = tuple([slice(self.pad,-self.pad)]
-                                *len(padded_buffer.shape))
-        else:
-            slice_pad = tuple([slice(None)]*len(padded_buffer.shape))
         # loop over the channels
-        mag2 = self.magnification**2
         for j in range(data.geometry.channels):
-            if channel_axis is not None:
-                slice_proj[channel_axis] = j
-            # loop over the projections
-            for i in tqdm(range(len(data.geometry.angles))):
-                
-                slice_proj[angle_axis] = i
-                padded_buffer.fill(0)
-                padded_buffer[slice_pad] = data.array[(tuple(slice_proj))]
+            # loop over the angles
+            for i in tqdm(range(len(data.geometry.angles)), disable = not self.verbose):
+                padded_buffer[:] = 0
+                padded_buffer[buffer_slice] = data.array[j, i, :, :]
 
                 if self.full_retrieval==True:
                     # apply the filter in fourier space, apply log and scale
-                    # by magnification
                     padded_buffer*=mag2
                     fI = fft2(padded_buffer)
                     iffI = ifft2(fI*self.filter).real
-                    np.log(iffI, out=padded_buffer)
-                    # apply scaling factor
-                    np.multiply(scaling_factor, padded_buffer, out=padded_buffer)
+                    padded_buffer[:] = scaling_factor*np.log(iffI)
+                    # (scaling_factor, padded_buffer, out=padded_buffer)
```

Comment:
why the change here? it's going to be a little less memory efficient than performing them in place.

Reply:
Good point, not sure why I changed it, put this back now